storage/sqlite/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+extern crate core;
+
+mod error;
+mod queries;
+
+use rusqlite::{Connection, OpenFlags, OptionalExtension};
+use std::borrow::Cow;
+use std::cmp::max;
+use std::path::PathBuf;
+use std::sync::{Mutex, MutexGuard};
+
+use crate::queries::SqliteQueries;
+use error::process_sqlite_error;
+use storage_core::{
+    backend::{self, TransactionalRo, TransactionalRw},
+    info::DbDesc,
+    Data, DbIndex,
+};
+use utils::shallow_clone::ShallowClone;
+use utils::sync::Arc;
+
+/// Sqlite iterator over entries with given key prefix
+pub struct PrefixIter {
+    /// Underlying iterator
+    iter: std::vec::IntoIter<(Vec<u8>, Vec<u8>)>,
+
+    /// Prefix to iterate over
+    prefix: Data,
+}
+
+impl PrefixIter {
+    fn new(iter: std::vec::IntoIter<(Vec<u8>, Vec<u8>)>, prefix: Data) -> Self {
+        PrefixIter { iter, prefix }
+    }
+}
+
+impl Iterator for PrefixIter {
+    type Item = (Data, Data);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let kv = self.iter.next()?;
+        utils::ensure!(kv.0.starts_with(&self.prefix));
+        Some(kv)
+    }
+}
+
+pub struct DbTx<'m> {
+    connection: MutexGuard<'m, Connection>,
+    queries: &'m SqliteQueries,
+}
+
+impl<'m> DbTx<'m> {
+    fn start_transaction(sqlite: &'m SqliteImpl) -> storage_core::Result<Self> {
+        let connection: MutexGuard<Connection> = sqlite
+            .connection
+            .lock()
+            .map_err(|_| storage_core::error::Recoverable::TemporarilyUnavailable)?;
```

Comment:
I think this is incorrect. Failing to lock a mutex is not a recoverable error. It should either block or there's a mutex poisoning which is practically a dead-lock.

If you want this to be recoverable, then you should use `try_lock()` instead of `lock()`, and then see what the error is.

Reply:
You are right, I changed the error to Fatal.